dbparser: add XML validation in test_patterndb

bazsi opened the pull request:
This series of patches improve test_patterndb test program a lot and also contains a number of refactors, in preparation for the new create-context action in patterndb.

It depends on PR #1004, so that should go in first, but I wanted to submit this separately to facilitate review in smaller batches.

bazsi commented:
I have now rebased this to master, so this should be the next step of integration in my dbparser test refactor series. Please don't merge anything to dbparser until this series of PRs are in, as this is pretty fragile, moving a lot of code around, thus even a single line changes in dbparser makes rebasing of this pretty difficult.

MrAnno commented on modules/dbparser/pdb-file.c:
```diff
@@ -0,0 +1,165 @@
+  fclose(pdb);
+  if (!result)
+    {
+      g_set_error(error, PDB_ERROR, PDB_ERROR_FAILED, "Error detecting pdbfile version, <patterndb> version attribute not found or <patterndb> is not on its own line");
```
Is this error message still relevant after `dbparser: improve pdb_file_detect_version in pdb-file`?

bazsi replied:
yup, it still has to be on the same line.

MrAnno replied:
Oh sorry, I misunderstood the message.

bkil-syslogng replied:
A bit narrower lines please?

MrAnno commented:
:+1:

bazsi commented:
Can I merge this with :+1: that @MrAnno posted?

bkil-syslogng commented on modules/dbparser/pdbtool/pdbtool.c:
```diff
@@ -707,20 +707,12 @@ pdbtool_test(int argc, char *argv[])
-          if (system(cmd) != 0)
-            {
-              fprintf(stderr, "%s: xmllint returned an error, the executed command was: %s", argv[arg_pos], cmd);
+              fprintf(stderr, "%s: error validating pdb file: %s\n", argv[arg_pos], error->message);
```
I liked the error==NULL check above. Want to do that here as well?

bkil-syslogng commented on modules/dbparser/pdb-file.c:
```diff
@@ -76,3 +78,88 @@ pdb_file_detect_version(const gchar *pdbfile, GError **error)
+      return FALSE;
+    }
+
+  xmllint_cmdline = g_strdup_printf("xmllint --noout --nonet --schema %s %s", xsd_file, filename);
```
Should we use quoting to help support space in installation folder/file names?

bkil-syslogng commented on modules/dbparser/pdb-file.c:
```diff
@@ -76,3 +78,88 @@ pdb_file_detect_version(const gchar *pdbfile, GError **error)
+static const gchar *
+_get_xsddir(void)
+{
+  return _get_xsddir_in_build() ? : _get_xsddir_in_production();
```
This works fine, although I would probably have `#ifdef`'ed the whole function definition above and this condition as well.

bkil-syslogng commented:
I didn't see any major issue with the code. I haven't verified the XML file changes, though. I assume tests work and run correctly. So other than the nitpicks above, :+1:

bazsi commented:
Thanks for the reviews. I have fixed two nitpicks:

1) too long line in pdb_file_detect_version()
2) and quoting in the invocation of xmllint

Note: that both were copies of code I've moved around and not new code.

I kept the conditionals around get_xsddir() as they were, I think it documents the intent pretty clearly: we use in-tree XSD files in case we are working in tree and installed versions if we are outside. If I was to add a lot of ifdefs there, that would be more difficult to read.

As these were only nitpicks that I mostly fixed, I would continue by merging this, once travis comes back green.

And after that's done, we should continue with PR #1012

lbudai commented on modules/dbparser/pdb-file.c:
```diff
@@ -77,3 +79,88 @@ pdb_file_detect_version(const gchar *pdbfile, GError **error)
+{
+#ifdef SYSLOG_NG_ENABLE_DEBUG
+  const gchar *srcdir;
+  static gchar path[256];
```
A small note here:
Maybe we could check if `PATH_MAX` is available or not,  if not then we could try to use pathconf (a separate `build_path(base_dir, relative_path)` function could be also added).

(IIRC I have a patch somewhere for determining path max ... maybe in PE).

This is not a blocker issue, it can be merged as-is (and later I'll send a patch).

bkil commented on modules/dbparser/pdb-file.c:
```diff
@@ -0,0 +1,166 @@
+  srcdir = getenv("top_srcdir");
+  if (srcdir)
+    {
+      g_snprintf(path, sizeof(path), "%s/doc/xsd", srcdir);
```
We will need to check the return value of `g_snprintf()` as well. Would some kind of `g_strdup_printf()` work here instead?

bazsi replied:
Hi,

The problem with allocating the string is that it makes calling this
function a lot more difficult, as the caller needs to free it.

This kind of pattern happens in syslog-ng a lot (e.g. formatting a string I
a static buffer and then returning that), and I would change them to use a
different pattern at once and not individually.

My point is that the caller shouldnt need to bother freeing the string as
that complicatedly call sites a lot. If we find a way how to do that, with
truncation and memory mgmt covered I am all for it.

BTW: truncation is not really a problem in this case but might in others.
On May 2, 2016 11:51 PM, "Thomas Nagy" <notifications@github.com> wrote:

> In modules/dbparser/pdb-file.c
> <https://github.com/balabit/syslog-ng/pull/1007#discussion_r61811173>:
>
> > +                  "Error detecting pdbfile version, <patterndb> version attribute not found or <patterndb> is not on its own line");
> > +    }
> > +  return result;
> > +}
> > +
> > +static const gchar *
> > +_get_xsddir_in_build(void)
> > +{
> > +#ifdef SYSLOG_NG_ENABLE_DEBUG
> > +  const gchar *srcdir;
> > +  static gchar path[256];
> > +
> > +  srcdir = getenv("top_srcdir");
> > +  if (srcdir)
> > +    {
> > +      g_snprintf(path, sizeof(path), "%s/doc/xsd", srcdir);
>
> We will need to check the return value of g_snprintf() as well. Would
> some kind of g_strdup_printf() work here instead?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/1007/files/9303b028cb03709baa83399624e236e7a808631c#r61811173>
>

bazsi commented:
I am merging this now, the snprintf() case should be handled as a separate issue, I have even filed an issue #1038
